keps/sig-scheduling/4671-gang-scheduling/README.md

```diff
@@ -0,0 +1,927 @@
+    // and referred to by the key's value in events and errors.
+    ReplicatedGangKey *ObjectFieldSelector
+
+    // Optional when GangMode = "ReplicatedGang".
+    // Forbidden otherwise.
+    Replicas int
+
+    // PodGroupSelector must be set when GangMode = "Single" or "Off".
+    // Must be one of:
+    //   - `metadata.labels['<KEY>']`
+    //   Note: Annotation not supported because it is not indexed easily, and it may not
+    //         have a form that can be printed without escaping.
+    // Each value of that key maps to a different gang, which will be anonymous in status,
+    // and referred to by the key's value in events and errors.
+    PodGroupSelector PodGroupSelector
+
+    // GangSchedulingPolicy defines the options applying to all pods in this gang.
+    // Forbidden if GangMode is set to "Off".
+    GangSchedulingPolicy GangSchedulingPolicy
+}
+
+
+type WorkloadStatus struct {
+  // Necessary status fields TBD.
+}
+```
+
+The individual `PodGroups` and `PodGroup` replicas are treated as independent gangs. As an example, if one of
+the groups can be scheduled and the other can't be - this is exactly what will happen. If the underlying
+user intention was to have either both of them or none of them running, they should form a single group and
+not be split into two. A `LeaderWorkerSet` is a good example of it, where a single `PodGroup` replica consists
+of a single leader and `N` workers and that forms a scheduling (and runtime unit), but workload as a whole
+may consist of a number of such replicas.
+
+### Scheduler Changes
+
+The kube-scheduler will be watching for `Workload` objects (using informers) and will use them to map pods
+to and from their `Workload` objects.
+
+In the initial implementation, we expect users to create the `Workload` objects. In the next steps controllers
+will be updated to create an appropriate `Workload` objects themselves whenever they can appropriately infer
+the intention from the desired state.
+Note that given scheduling options are stored in the `Workload` object, pods linked to the `Workload`
+object will not be scheduled until this `Workload` object is created and observed by the kube-scheduler.
+
+The north star vision for gang scheduling implementation should satisfy the following requirements:
+
+1. Ensure that pods being part of a gang are not scheduled if all pods belonging to it can't be scheduled.
+2. Provide the "optimal enough" placement by considering all pods from a gang together.
+3. Avoid deadlock scenario when multiple workloads are being scheduled at the same time by kube-scheduler.
+4. Avoid deadlock scenario when multiple workloads are being scheduled at the same time by different
+   schedulers.
+5. Avoid preemptions of already running pods if these will not allow to schedule higher priority gang
+   anyway.
+6. Support gang-level (or workload-level in general) level preemption.
```

Comment:
Can we clarify a bit more? When I say "gang-aware preemption", I mean all pods from the group should be schedulable after the preemption. i.e., if pod#1-pod#15 can be schedulable after the preemption, but if the preemption cannot help making pod#16 schedulable, then all preemption shouldn't be performed.

I'd like to make sure we're on the same page, or there could be more basic requirements.

Reply:
> When I say "gang-aware preemption", I mean all pods from the group should be schedulable after the preemption. i.e., if pod#1-pod#15 can be schedulable after the preemption, but if the preemption cannot help making pod#16 schedulable, then all preemption shouldn't be performed.

Right - that's exactly what (5) was supposed to say. Hopefully the clarification suggested by Dominik helps here.

Regarding (6) - this is effectively a supplement of (5):
- in (5) we're saying that we won't preempt something, if we won't be schedule a gang anyway
- in (6) we're saying that if pods form a gang also from a runtime perspective, then we won't be preempting individual pods, but rather will be thinking about these as units

Reply:
> in (6) we're saying that if pods form a gang also from a runtime perspective, then we won't be preempting individual pods, but rather will be thinking about these as units

Ok - please elaborate the explanation then. It's not very clear now

Reply:
Expanded the point (6) - PTAL